Participant metering documentation (#13070)
digital-asset/daml: docs/source/ops/metering.rst

.. Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Participant Metering
====================

Participant metering is a way to report how many events have been submitted in a given period of time.

Daml command execution results in a Daml transaction that contains events associated with the processing of the command.

The events included in the report include:

    * Contract creation
    * Exercise of a contract (including non-consuming exercises and exercise by key)
    * Fetch of a contract (including fetch by key)
    * Lookup by contract key

Only events that originated from the local participant are included in the metering.  Events received
by the local participant from remote participants are *not* included.

> Only events contained in an accepted transactions (ones that maintains the
> :doc:`integrity of the ledger model </concepts/ledger-model/ledger-integrity>`) are included.

Reviewer: ```suggestion
Only events contained in committed transactions are included, a failed transaction has no effect on ledger metering.
```

.. Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Participant Metering
====================

Participant metering is a way to report how many events have been submitted in a given period of time.

Daml command execution results in a Daml transaction that contains events associated with the processing of the command.

The events included in the report include:

>     *  Contract creation
>     *  Consuming exercise by contract identifier
>     *  Non-consuming exercise by contract identifier
>     *  Consuming contract exercise by contract key
>     *  Non-consuming contract exercise by contract key
>     *  Contract fetch by contract identifier
>     *  Contract fetch by contract key
>     *  Contract lookup by contract key

Reviewer: ```suggestion
    *  Contract creation
    *  Exercise of a contract (including non-consuming exercises and exercise by key)
     * Fetch of a contract (including fetch by key)
     * Lookup by contract key
```
I think that makes it a bit clearer that this just corresponds to the four type of transaction nodes we have.

> Note that only events that originated from the local participant are included in the metering.  Events received

Reviewer: It might also be worth pointing out exlicitly that only a successful, committed transaction is metered not one that fails. Saying you meter tranascitons rather than commands points that out implicitly but I think it’s worth making explicit.

Author: Done.

by the local participant from remote participants are *not* included.

Generating a Metering Report
----------------------------

A metering report is generated using the :doc:`Daml assistant </tools/assistant>` utility.

To run a metering report ``daml ledger metering-report`` is used with the following metering specific arguments:

> :--from:

Reviewer: ```suggestion
:``--from``:
```

Reviewer: otherwise the double dashes are not rendered as double dashes.

>     The period being reported on will start from the given date. Events on or after this date will be included.

Reviewer 2: ```suggestion
    A start date that is used to initiate the reporting period. Events on or after this date will be included.
```
To create parallel construction with the end date description.

Author: Done.

> :--to:

Reviewer: ```suggestion
:``--to``:
```

    An end date that may be used to terminate the reporting period.  Events prior to this date will be included.

>     If an end date is not provided then the report will contain counts of all events that occurred after the to

Reviewer 2: Do you mean "after the from date"?

Author: Yes.

    the ``--from`` date.

> :--application:

Reviewer: ```suggestion
:``--application``:
```

>     An optional application can be provided to limit the report to a single application.

Reviewer 2: ```suggestion
    Optionally, provide an application to limit the report to that application.
```

Author: Done.

> The from and to dates above should be provided formatted as ``yyyy-mm-dd``.  The exact timestamp used for the report

Reviewer 2: ```suggestion
The from and to dates above should be formatted ``yyyy-mm-dd``.  The exact timestamp used for the report
```

Author: Done.

.. Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Participant Metering
====================

Participant metering is a way to report how many events have been submitted in a given period of time.

> This may form part of the commercial arrangement Digital Asset have with customers.

Author: Done.

The events included in the report include:

    *  Contract creation
    *  Consuming exercise by contract identifier
    *  Non-consuming exercise by contract identifier
    *  Consuming contract exercise by contract key
    *  Non-consuming contract exercise by contract key
    *  Contract fetch by contract identifier
    *  Contract fetch by contract key
    *  Contract lookup by contract key

Note that only events that originated from the local participant are included in the metering.  Events received
by the local participant from remote participants are *not* included.

Generating a Metering Report
----------------------------

A metering report is generated using the :doc:`Daml assistant </tools/assistant>` utility.

To run a metering report ``daml ledger metering-report`` is used with the following metering specific arguments:

:--from:
    The period being reported on will start from the given date. Events on or after this date will be included.

:--to:
    An end date that may be used to terminate the reporting period.  Events prior to this date will be included.

>     If an end date is not provided then the report will contains counts of all events that occurred after the to

Author: Done

will be the start of the UTC day provided.

Other non-metering specific Daml assistant flags may also be used alongside those shown above.

Example
-------

To report on all applications for January 2022 the following from/to setting would be used:

``daml ledger metering-report --from 2022-01-01 --to 2022-02-01``

Output
------

.. code-block:: json

    {
        "participant": "some-participant",
        "request": {
            "from": "2022-01-01T00:00:00Z",
            "to": "2022-02-01T00:00:00Z"
        },
        "final": true,
        "applications": [
            {
                "application": "some-application",
                "events": 42
            }
        ]
    }

The output consists of the following sections:

:participant:
    The name of the local participant the report applies to

:request:
    This section gives details of the parameters that were used to generate the report

:final:
    This field will be set to ``true`` if a from date was provided and the from date is
    in the past.  Once a report is marked as final the event counts will never change so

>     may be used for billing purposes.

Author: Good spot, `from` should be `to` !
